etl/storage/storage.go

```diff
@@ -82,6 +85,105 @@ func (s *storage) GetRecord(ref models.Reference) (models.Record, error) {
 	return record, err
 }
 
+func checkIndex(i string) (int, int, error) {
+	index := strings.Split(i, ":")
+	if len(index) != 2 {
+		return 0, 0, errors.New("query parameter 'index' should have the '<pulse_number>:<order>' format")
+	}
+	var err error
+	var pulseNumber, order int64
+	pulseNumber, err = strconv.ParseInt(index[0], 10, 64)
+	if err != nil {
+		return 0, 0, errors.New("query parameter 'index' should have the '<pulse_number>:<order>' format")
+	}
+	order, err = strconv.ParseInt(index[1], 10, 64)
+	if err != nil {
+		return 0, 0, errors.New("query parameter 'index' should have the '<pulse_number>:<order>' format")
+	}
+	return int(pulseNumber), int(order), nil
+}
+
+func filterByPulse(query *gorm.DB, pulseNumberLt, pulseNumberGt *int) *gorm.DB {
+	if pulseNumberGt != nil {
+		query = query.Where("pulse_number > ?", *pulseNumberGt)
+	}
+	if pulseNumberLt != nil {
+		query = query.Where("pulse_number < ?", *pulseNumberLt)
+	}
+	return query
+}
+
+func filterRecordsByIndex(query *gorm.DB, fromIndex string, sort string) (*gorm.DB, error) {
+	pulseNumber, order, err := checkIndex(fromIndex)
+	if err != nil {
+		return nil, err
+	}
+	switch sort {
+	case "asc":
+		query = query.Where("(pulse_number > ?", pulseNumber)
+		query = query.Or("pulse_number = ? AND \"order\" >= ?)", pulseNumber, order)
+	case "desc":
+		query = query.Where("(pulse_number < ?", pulseNumber)
+		query = query.Or("pulse_number = ? AND \"order\" <= ?)", pulseNumber, order)
+
+	default:
+		return nil, errors.New("query parameter 'sort' should be 'asc'' or 'desc'")
+	}
+	return query, nil
+}
+
+func sortRecordsByDirection(query *gorm.DB, sort string) (*gorm.DB, error) {
+	switch sort {
+	case "asc":
+		query = query.Order("pulse_number asc").Order("\"order\" asc")
+	case "desc":
+		query = query.Order("pulse_number desc").Order("\"order\" desc")
+	default:
+		return nil, errors.New("query parameter 'sort' should be 'asc'' or 'desc'")
+	}
+	return query, nil
+}
+
+func getRecords(query *gorm.DB, limit, offset int) ([]models.Record, int, error) {
+	records := []models.Record{}
+	var total int
+	err := query.Limit(limit).Offset(offset).Find(&records).Error
+	if err != nil {
+		return nil, 0, err
+	}
+	err = query.Count(&total).Error
+	if err != nil {
+		return nil, 0, err
+	}
+	return records, total, nil
+}
+
+// GetLifeline returns records for provided object reference, ordered by pulse number and order fields.
+func (s *storage) GetLifeline(objRef []byte, fromIndex *string, pulseNumberLt, pulseNumberGt *int, limit, offset int, sort string) ([]models.Record, int, error) {
```

Comment:
why did you pass a pointer of string and int? seems that will not change inside the function.  if changes are not expected just delete it.

Reply:
I check if pointers of int are nil in function, called inside this func: `filterByPulse`. And I check `fromIndex` param to decide to call `filterRecordsByIndex` func or not